Docs: Add Next.js framework doc

kylegach opened the pull request:
## What I did

Documents Storybook for Next.js framework

Based on
- https://github.com/storybookjs/storybook/tree/next/code/frameworks/nextjs
- https://storybook.js.org/blog/integrate-nextjs-and-storybook-automatically/

![](https://github.com/storybookjs/storybook/assets/486540/afea619e-5361-48ae-878c-2c0554399c44)

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Follow the steps in the [contributing instructions](https://storybook.js.org/docs/react/contribute/new-snippets#preview-your-work) for this branch, `framework-doc-nextjs`

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [x] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

kylegach committed:
Add Next.js framework doc

- Based on
    - https://github.com/storybookjs/storybook/tree/next/code/frameworks/nextjs
    - https://storybook.js.org/blog/integrate-nextjs-and-storybook-automatically/

valentinpalkovic commented on docs/snippets/react/nextjs-add-framework.js.mdx:
```diff
@@ -0,0 +1,16 @@
+    name: '@storybook/nextjs', // 👈 Add this
+    options: {
+      builder: {
+        // Set useSWC to true if you want to try out the experimental SWC compiler in Next.js >= 14.0.0
+        useSWC: true,
+      },
```
`useSWC` does not exist in Storybook 8.0 anymore. See [the migration doc].(https://github.com/storybookjs/storybook/blob/valentin/remove-babel-from-webpack5-builder/MIGRATION.md#automatic-swc-mode-detection)

valentinpalkovic commented on docs/snippets/react/nextjs-add-framework.ts.mdx:
```diff
@@ -0,0 +1,20 @@
+  framework: {
+    // name: '@storybook/react-webpack5', // 👈 Remove this
+    name: '@storybook/nextjs', // 👈 Add this
+    options: {
+      builder: {
+        // Set useSWC to true if you want to try out the experimental SWC compiler in Next.js >= 14.0.0
+        useSWC: true,
+      },
+    },
```
`useSWC` does not exist in Storybook 8.0 anymore. See [the migration doc](https://github.com/storybookjs/storybook/blob/valentin/remove-babel-from-webpack5-builder/MIGRATION.md#automatic-swc-mode-detection).

kylegach committed:
Address comments

- Remove references to `useSWC`

kylegach committed:
Update RSC API

- Per https://github.com/storybookjs/storybook/pull/25591
- `features.experimentalNextRSC` -> `features.experimentalRSC`
- `parameters.nextjs.rsc` -> `parameters.react.rsc`

kylegach committed:
Update README to point to docs

chantastic reviewed (approved):
Love seeing this. That's a lot to cover, nicely done.

jonniebigodes commented on docs/get-started/nextjs.md:
```diff
@@ -0,0 +1,921 @@
+
+## Postcss
+
+Next.js lets you [customize postcss config](https://nextjs.org/docs/pages/building-your-application/configuring/post-css). Thus this framework will automatically handle your postcss config for you.
+
+This allows for cool things like zero config tailwindcss! (See [Next.js' example](https://github.com/vercel/next.js/tree/canary/examples/with-tailwindcss))
```
When you can, can you capitalize PostCSS and adjust Tailwind's naming, as we're not referring to the package but the brand?

kylegach replied:
Good catch! Same goes for SVGR, Webpack, Babel, etc.

jonniebigodes commented on docs/get-started/nextjs.md:
````diff
@@ -0,0 +1,921 @@
+
+[Absolute imports](https://nextjs.org/docs/pages/building-your-application/configuring/absolute-imports-and-module-aliases#absolute-imports) from the root directory are supported.
+
+```jsx
````
When you're able can you add a comment referencing the filename so that it depicts the filename accurately? Optionally, this could be converted into two examples one for JavaScript and one for TypeScript

jonniebigodes commented on docs/get-started/nextjs.md:
```diff
@@ -0,0 +1,921 @@
+
+## Custom Webpack Config
+
+Next.js comes with a lot of things for free out of the box like sass support, but sometimes you add [custom webpack config modifications to Next.js](https://nextjs.org/docs/pages/api-reference/next-config-js/webpack). This framework takes care of most of the webpack modifications you would want to add. If Next.js supports a feature out of the box, then that feature will work out of the box in Sto ...
```
Can you check the capitalization of Sas, Webpack, and SVGR so that we can ensure it's accurate?

jonniebigodes commented on docs/snippets/react/nextjs-configure-svgr.js.mdx:
````diff
@@ -0,0 +1,25 @@
+```js
````
Can you adjust this file and the TypeScript variant to align with our formatting pattern?

jonniebigodes commented on docs/snippets/react/rsc-feature-flag.js.mdx:
````diff
@@ -0,0 +1,9 @@
+```js
````
Once again, the same formatting should be applied here and also for the associated variant.

jonniebigodes commented on docs/get-started/nextjs.md:
```diff
@@ -0,0 +1,921 @@
+
+<!-- prettier-ignore-end -->
+
+Note that wrapping your server components in Suspense does not help if your server components access server-side resources like the file system or Node-specific libraries. To deal work around this, you'll need to mock out your data access layer using [Webpack aliases](https://webpack.js.org/configuration/resolve/#resolvealias) or an addon like [storybook-addon-module-mock](https://storybook.js.or ...
```
Correct me if I'm wrong, but am I missing something here, or is this "To deal work around" a typo? Shouldn't this be something like "To deal with...", "To workaround this", or something of that nature?

jonniebigodes commented on docs/get-started/nextjs.md:
````diff
@@ -0,0 +1,921 @@
+}
+```
+
+Therefore, if something in storybook isn't showing the image properly, make sure you expect the object to be returned from an import instead of only the asset path.
````
When you're able, can you capitalize Storybook for accuracy?

jonniebigodes commented on docs/get-started/nextjs.md:
```diff
@@ -0,0 +1,921 @@
+
+The absolute path to the `next.config.js` file. This is necessary if you have a custom `next.config.js` file that is not in the root directory of your project.
+
+## Acknowledgements
```
Can you adjust this and move the Acknowledgements back into the README? I'm grateful for the fantastic work the addon authors did by creating, maintaining, and spearheading the Next.Js integration, but in my opinion, this should be referenced elsewhere.

kylegach replied:
Yeah, that's a fair point. Done.

jonniebigodes reviewed (changes requested):
I took a preliminary pass at this and there's some work that needs to done for now so that we can get this into a good shape before we can take another pass at it and polish it up in terms of structure.

kylegach commented on docs/snippets/react/nextjs-add-framework.js.mdx:
````diff
@@ -0,0 +1,8 @@
+```js
+// .storybook/main.js
````
Now that the first line comment is extracted into the snippet title, that line break is no longer necessary.

jonniebigodes replied:
Understood, then if you're ok with it adjust the remainder examples to align to this pattern instead

kylegach commented on docs/snippets/react/nextjs-navigation-segments-for-use-param-override-in-meta.js.mdx:
````diff
@@ -0,0 +1,31 @@
+```js
````
I tried to do that first, but our CodeSnippets logic doesn't handle it well and only rendered one of the snippets instead of both tabs.

kylegach replied:
I think I found a nice way of handling this:

<img width="698" alt="" src="https://github.com/storybookjs/storybook/assets/486540/9ab5b0d3-7ab1-4326-91f2-ea7555336892">

Apologies if this is what you were actually suggesting. 🙃

jonniebigodes replied:
No need to apologize whatsoever; that's what I was aiming for 🙂!

kylegach commented on docs/snippets/react/nextjs-router-override-in-story.js.mdx:
````diff
@@ -0,0 +1,24 @@
+```js
+// SomeComponentThatUsesTheRouter.stories.js
+import SomeComponentThatUsesTheRouter from './SomeComponentThatUsesTheRouter';
````
The `Router` and `Navigation` parts are key words that pertain to the relevant Next.js API and should be maintained.

But I can shorten it to `RouterBasedComponent` and `NavigationBasedComponent`.

kylegach commented on docs/get-started/nextjs.md:
```diff
@@ -0,0 +1,921 @@
+
+<!-- prettier-ignore-end -->
+
+The parameter `nextjs.appDirectory` defaults to `false` if not set.
```
Honestly, an unset property defaulting to false probably doesn't even need to be documented. For completeness's sake though, I opted to document this in the API section, as the default value.

kylegach commented on docs/get-started/nextjs.md:
````diff
@@ -0,0 +1,921 @@
+
+[Local images](https://nextjs.org/docs/pages/building-your-application/optimizing/images#local-images) are supported.
+
+```jsx
````
This is demonstrating usage of a Next.js feature, not a Storybook one. This is not something folks will be copy/pasting, and if they do, it'll be the `<Image>` portion, which will be identical for JS & TS. I don't think the overhead of multiple snippets is worth the small benefit.

kylegach commented on docs/get-started/nextjs.md:
````diff
@@ -0,0 +1,921 @@
+
+[css modules](https://nextjs.org/docs/pages/building-your-application/styling/css-modules) work as expected.
+
+```jsx
````
This snippet only illustrates that imports function correctly. Making a TS version would be unnecessary overhead (and would also be virtually identical, with only the filename changing).

kylegach committed:
Address comments

- Add missing filenames or titles to snippets
- Proper capitalization of terms like Webpack, Babel, PostCSS, etc.
- Fix some typos
- Split up complex snippets
- Document default parameter values
- Move acknowledgements back to README

kylegach commented:
@jonniebigodes — Ready for your re-review!

valentinpalkovic commented on docs/snippets/react/nextjs-navigation-push-override-in-preview.js.mdx:
```diff
@@ -0,0 +1,21 @@
+    // ...
+    nextjs: {
+      router: {
+        navigation(...args) {
+          // Custom logic can go here
+          // This logs to the Actions panel
+          action('nextNavigation.push')(...args);
+          // Return whatever you want here
+          return Promise.resolve(true);
+        },
+      },
```
This seems to be wrong. Please take a look at the Typescript variant

kylegach committed:
Address comments

- Simplify navigation/router fn override guidance
- Fix incorrect snippet
- Consistently format headings
